### PR TITLE
Draw fish bones for Path3D and Path2D in the Editor

### PR DESCRIPTION
**Pull request by xiongyaohua**
Visualize `Path3D` direction and tilt with the so called `fish bones`.

image

For `Path2D`

Screenshot 2022-11-19 at 15 21 47

For more details, see  https://github.com/godotengine/godot-proposals/issues/5716 and https://github.com/godotengine/godot-proposals/issues/5811

### CODE REVIEW

File: editor/plugins/path_3d_editor_plugin.cpp [hunk 1]
```diff
@@ -239,38 +239,69 @@ void Path3DGizmo::redraw() {
 		return;
 	}
 
-	Vector<Vector3> v3a = c->tessellate();
+	//Vector<Vector3> v3a = c->tessellate();
 	//Vector<Vector3> v3a=c->get_baked_points();
```
Comment (fire) [inline editor/plugins/path_3d_editor_plugin.cpp:243]: Prefer to delete code in comments.

**Comment by TokageItLab** [comment]
See also #68873

File: editor/plugins/path_3d_editor_plugin.cpp (hunk 1 above)
Comment (xiongyaohua) [reply editor/plugins/path_3d_editor_plugin.cpp]: deleted

**Comment by xiongyaohua** [comment]
> See also https://github.com/godotengine/godot/pull/68873

I think that PR is a good complement to this one. Even better, this PR didn't touch any code related with gizmo handles, so should be easy to merge.

**Comment by xiongyaohua** [comment]
Change the visual of 3D bones, to keep it consistent with 2D bones. Since the 2D bones are facing left and right, the 3D bones do the same. But in 3D we also need to indicate the up direction, so smaller bones pointing up are added.

Screenshot 2022-11-20 at 07 58 02

**Review by mhilbrunner — APPROVED** [review]
Feature approved in today's PR meeting, needs review by @godotengine/3d-editor

File: scene/2d/path_2d.cpp [hunk 2]
```diff
@@ -106,18 +106,58 @@ void Path2D::_notification(int p_what) {
...
+				const Transform2D *r = frames.ptr();
+				// Draw curve segments
+				{
+					PackedVector2Array v2p;
+					v2p.resize(sample_count);
+					Vector2 *w = v2p.ptrw();
+
+					for (int i = 0; i < sample_count; i++) {
+						w[i] = r[i].get_origin();
+						;
```
[the remark this answers is not in the document]
Comment (xiongyaohua) [reply scene/2d/path_2d.cpp:136]: removed

File: editor/plugins/path_3d_editor_plugin.cpp [hunk 3]
```diff
@@ -239,38 +239,65 @@ void Path3DGizmo::redraw() {
...
+		{
+			Transform3D *w = frames.ptrw();
+
+			for (int i = 0; i < sample_count; i++) {
+				w[i] = c->sample_baked_with_rotation(i * interval, true, true);
+			}
+		}
+
+		// Is this BUG still relevant? Seems not.
+		// BUG: the following won't work when v3s, avoid drawing as a temporary workaround.
```
[the remark this answers is not in the document]
Comment (xiongyaohua) [reply editor/plugins/path_3d_editor_plugin.cpp:262]: removed

**Comment by xiongyaohua** [comment]
Small Tweak of 2D bones

new
image

old
image

**Comment by xiongyaohua** [comment]
squashed.

**Comment by xiongyaohua** [comment]
> Feature approved in today's PR meeting, needs review by https://github.com/orgs/godotengine/teams/3d-editor

Hi, is there only progress on this front? If there is no major problem is it possible to merge it first. I plan to do further refactor, improving the usability of `CSGPolygon3D`, which depends on this PR. Only minor problems or tweaks(if any) I can do it in followup PRs.

**Review by TokageItLab — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**